File: Commander/platforms/Windows/Root.cs
#if Windows
using AspNetExtensions;
using CsTools.Extensions;
using CsTools.Functional;

using static CsTools.Core;

record RootItem(
    string Name,
    string? Description,
    long? Size,
    bool IsMounted
)
{
    public static RootItem Create(DriveInfo driveInfo)
        => new(
            driveInfo.Name,
            driveInfo.IsReady ? driveInfo.VolumeLabel : null,
            driveInfo.IsReady ? driveInfo.TotalSize : null,
            driveInfo.IsReady);   
};

static class Root
{
    public static AsyncResult<RootItem[], RequestError> Get()
        =>  Ok<RootItem[], RequestError>((from n in DriveInfo
                        .GetDrives()
                        .Select(RootItem.Create)
                        .Append(new("zzz", "", 0, true))
            orderby n.IsMounted descending, n.Name
            select n.Name == "zzz" 
                ? n with { Name = "services" } 
                : n).ToArray())
            .ToAsync()
            .ToAsyncResult();
}

#endif

File: Windows/Root.fs
module Root
open System.IO
open Types
open RequestResult

type RootItem = {
    Name: string
    Description: string option
    Size: int64 option
    IsMounted: bool
}

let get (_: Empty) = 
    let createRootItem (driveInfo: DriveInfo) = 
        {
            Name = driveInfo.Name
            Description = if driveInfo.IsReady then Some driveInfo.VolumeLabel else None
            Size = if driveInfo.IsReady then Some driveInfo.TotalSize else None
            IsMounted = driveInfo.IsReady
        }

    let appendServiceItem items = 
        Seq.append items [| { Name = "services"; Description = None; Size = None; IsMounted = true } |]

    returnReqVal (DriveInfo.GetDrives ()
                    |> Seq.map createRootItem
                    |> Seq.sortBy (fun item -> (not item.IsMounted, item.Name))
                    |> appendServiceItem
                    |> Seq.toArray)


